### PR TITLE
Document Generator Small full balance sheet Period IXBRL model Mapper

### PR DESCRIPTION
**Pull request by Mathsh**
Using MapStruct created an ApiToPeriod mapper to map the api data for period to the IXBRL model for small full

- Updated SDK version
- Added depedancy for company-accounts-library to enable the use of the datesHelper
- Updated the maven compiler version
- Added the mapStruct dependancy
- Created ApiToPeriodMapper interface with the @Mapping details
- Created ApiToPeriodMapperDecorator and added the @DecoratedWith annotation to ApiToWebMapper

The decorator is to enable custom mapping outside of the mapStruct as we need to format some of the dates held in the Period model using the DatesHelper from company-accounts-library

- Created Junit tests to test mapper.

Resolves SFA 646 & 857

### CODE REVIEW

File: document-generator-accounts/pom.xml [hunk 1]
```diff
@@ -26,13 +26,15 @@
 
         <!--Dependencies-->
         <spring-boot-start-web.version>2.0.4.RELEASE</spring-boot-start-web.version>
-        <api-sdk-java.version>3.0.0-rc1</api-sdk-java.version>
+        <api-sdk-java.version>3.6.0-rc1</api-sdk-java.version>
```
Comment (ch-mcabrera) [inline document-generator-accounts/pom.xml]: I think the latest skd is `<api-sdk-java.version>3.8.0-rc1</api-sdk-java.version>`

File: document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapper.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    ApiToPeriodMapper INSTANCE = Mappers.getMapper(ApiToPeriodMapper.class);
+
+    @Mappings({
+            @Mapping(source = "accounts.nextAccounts.periodStartOn", target = "currentPeriodStartOn"),
+            @Mapping(source = "accounts.nextAccounts.periodEndOn", target = "currentPeriodEndsOn"),
+            @Mapping(source = "accounts.lastAccounts.periodStartOn", target = "previousPeriodStartOn"),
+            @Mapping(source = "accounts.lastAccounts.periodEndOn", target = "previousPeriodEndsOn"),
+    })
+    Period apiToPeriod(CompanyProfileApi companyProfile);
```
Comment (ch-mcabrera) [inline document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapper.java]: We could rename this method to be more descriptive e.g. `convertCompanyProfileApiToPeriod`

File: document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperDecorator.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @Override
+    public Period apiToPeriod(CompanyProfileApi companyProfile) {
+
+        Period period = apiToPeriodMapper.apiToPeriod(companyProfile);
+
+        boolean isSameYear = isSameYearFiler(companyProfile);
+
+        NextAccountsApi nextAccounts = companyProfile.getAccounts().getNextAccounts();
```
Comment (ch-mcabrera) [inline document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperDecorator.java]: what would happen if `getAccounts` return `null` or  `getNextAccounts` return `null`?

File: pom.xml [hunk 4]
```diff
@@ -28,7 +28,7 @@
         <jdk.version>1.8</jdk.version>
 
         <!-- Maven and Surefire plugins -->
-        <maven-compiler-plugin.version>3.1</maven-compiler-plugin.version>
+        <maven-compiler-plugin.version>3.5.1</maven-compiler-plugin.version>
```
Comment (ch-mcabrera) [inline pom.xml]: shouldnt this be `<maven-compiler-plugin.version>3.1</maven-compiler-plugin.version>`?

File: document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperDecorator.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    private boolean isSameYearFiler(CompanyProfileApi companyProfile) {
+        if (isMultipleYearFiler(companyProfile)) {
+            LastAccountsApi lastAccountsApi = companyProfile.getAccounts().getLastAccounts();
+            NextAccountsApi nextAccountsApi = companyProfile.getAccounts().getNextAccounts();
+            return accountsDatesHelper.isSameYear(lastAccountsApi.getPeriodEndOn(), nextAccountsApi.getPeriodEndOn());
+        }
+        return false;
+    }
+
+    private String convertToDisplayDate(LocalDate date) {
```
Comment (ch-mcabrera) [inline document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperDecorator.java]: We could add a comment here to clarify the fomat `YYYY MM dd` .

File: document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperDecorator.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        return accountsDatesHelper.convertLocalDateToDisplayDate(date);
+    }
+
+    private String convertDateToBSDate(LocalDate dateStartOn, LocalDate dateEndOn, Boolean isSameYear) {
```
Comment (ch-mcabrera) [inline document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperDecorator.java]: same here

File: document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperDecorator.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+        return period;
+    }
+
+    private boolean isMultipleYearFiler(CompanyProfileApi companyProfile) {
+        LastAccountsApi lastAccountsApi = companyProfile.getAccounts().getLastAccounts();
+        return lastAccountsApi != null && lastAccountsApi.getPeriodEndOn() != null;
+    }
+
+    private boolean isSameYearFiler(CompanyProfileApi companyProfile) {
```
Comment (ch-mcabrera) [inline document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperDecorator.java]: We could add a comment here to clarify what this does

File: document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperDecorator.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+        NextAccountsApi nextAccounts = companyProfile.getAccounts().getNextAccounts();
+        period.setCurrentPeriodStartOnFormatted(convertToDisplayDate(nextAccounts.getPeriodStartOn()));
+        period.setCurrentPeriodEndOnFormatted(convertToDisplayDate(nextAccounts.getPeriodEndOn()));
+        period.setCurrentPeriodBSDate(convertDateToBSDate(nextAccounts.getPeriodStartOn(),
+                nextAccounts.getPeriodEndOn(), isSameYear));
+
+
+        if (isMultipleYearFiler(companyProfile)) {
+            LastAccountsApi lastAccounts = companyProfile.getAccounts().getLastAccounts();
```
Comment (ch-mcabrera) [inline document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperDecorator.java]: we could move this to its own method `setPreviousPeriod(period, companyProfile)` same with the current period

File: document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapper.java (hunk 2 above)
Comment (Mathsh) [reply document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapper.java]: Think apiToPeriod within the context of /mapping/smallfull/mappers/ApiToPeriodMapper.java explains what is happening. Trying to keep the naming short and to the point.

File: document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperDecorator.java (hunk 3 above)
Comment (Mathsh) [reply document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperDecorator.java]: GetAccounts and GetNextAccounts would already have been checked before we hit mapping as you need current account data. Last account checks if it isMultipleYearFiler(companyProfile).

File: document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperDecorator.java (hunk 5 above)
Comment (Mathsh) [reply document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperDecorator.java]: A comment already exists on the dateHelper method it calls

File: document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperDecorator.java (hunk 6 above)
Comment (Mathsh) [reply document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperDecorator.java]: same as above

File: document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperDecorator.java (hunk 7 above)
Comment (Mathsh) [reply document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperDecorator.java]: It was discussed previously that if the naming of the method on a private method is self explanitory that we should not add javaDocs, and we should not be adding comments in java code. The method explains is same year filer.

File: document-generator-accounts/pom.xml (hunk 1 above)
Comment (Mathsh) [reply document-generator-accounts/pom.xml]: The SDK updates so quickly. I'll need to test again on the update, which I will do.

File: pom.xml (hunk 4 above)
Comment (Mathsh) [reply pom.xml]: 3.1 does not allow the use of annotationProcessorPaths so required 3.5.1

File: document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperDecorator.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+            period.setPreviousPeriodStartOnFormatted(convertToDisplayDate(lastAccounts.getPeriodStartOn()));
+            period.setPreviousPeriodEndOnFormatted(convertToDisplayDate(lastAccounts.getPeriodEndOn()));
+            period.setPreviousPeriodBSDate(convertDateToBSDate(lastAccounts.getPeriodStartOn(),
+                    lastAccounts.getPeriodEndOn(), isSameYear));
+        }
+    }
+
+    private boolean isMultipleYearFiler(CompanyProfileApi companyProfile) {
+        LastAccountsApi lastAccountsApi = companyProfile.getAccounts().getLastAccounts();
+        return lastAccountsApi != null && lastAccountsApi.getPeriodEndOn() != null;
```
Comment (ch-mcabrera) [inline document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperDecorator.java]: Just wondering, we are checking ` lastAccounts.getPeriodEndOn() != null`. 
Wouldn't we need to check `lastAccounts.getPeriodStartOn() != null`? Considering we then passed it to `convertToDisplayDate(lastAccounts.getPeriodStartOn())`.

File: document-generator-accounts/src/test/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperTest.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+        assertEquals(CURRENT_PERIOD_END_ON_FORMATTED, period.getCurrentPeriodEndOnFormatted());
+        assertEquals(PREVIOUS_PERIOD_START_ON_FORMATTED, period.getPreviousPeriodStartOnFormatted());
+        assertEquals(PREVIOUS_PERIOD_END_ON_FORMATTED, period.getPreviousPeriodEndOnFormatted());
+        assertEquals(CURRENT_PERIOD_BS_DATE, period.getCurrentPeriodBSDate());
+        assertEquals(PREVIOUS_PERIOD_BS_DATE, period.getPreviousPeriodBSDate());
+    }
+
+    @Test
+    @DisplayName("tests that the dates from Api are mapped to Period IXBRL model for single year filings")
+    public void testApiMapsDatesToPeriodModelForSingleYearFiling() {
```
Comment (ch-mcabrera) [inline document-generator-accounts/src/test/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperTest.java]: I would remove the `public` modifier as it is not needed in Junit 5

File: document-generator-accounts/src/test/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperTest.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+@ExtendWith(MockitoExtension.class)
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+public class ApiToPeriodMapperTest {
+
+    @Autowired
+    private ApiToPeriodMapper apiToPeriodMapper;
```
Comment (ch-mcabrera) [inline document-generator-accounts/src/test/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperTest.java]: Shouldnt this be `@InjectMocks`?

File: document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperDecorator.java (hunk 9 above)
Comment (Mathsh) [reply document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperDecorator.java]: we are checking if it is a multiple year filer. So we check if there is a lastAccountApi and if that has ended with GetPeriodEndOn. Else it is not a multi year file.

File: document-generator-accounts/src/test/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperTest.java (hunk 11 above)
Comment (Mathsh) [reply document-generator-accounts/src/test/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperTest.java]: it shouldn't be there at all. So I have removed

File: document-generator-accounts/src/test/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperTest.java (hunk 10 above)
Comment (Mathsh) [reply document-generator-accounts/src/test/java/uk/gov/companieshouse/document/generator/accounts/mapping/smallfull/mappers/ApiToPeriodMapperTest.java]: done

**Pull request merged**